### PR TITLE
Fix box coder op

### PR DESCRIPTION
**Pull request by Noplz**
Fix box decoder op

### CODE REVIEW

File: paddle/fluid/operators/box_coder_op.cc [hunk 1]
```diff
@@ -70,7 +77,7 @@ class BoxCoderOpMaker : public framework::OpProtoAndCheckerMaker {
              "of variance.");
     AddInput(
         "TargetBox",
-        "(LoDTensor or Tensor) this input is a 2-D LoDTensor with shape "
+        "(LoDTensor or Tensor) this input is a 2-D/3-D LoDTensor with shape "
```
Comment (qingqing01) [inline paddle/fluid/operators/box_coder_op.cc]: This input can be a 2-D LoDTensor with shape [N, 4] when `code_type` is 'encode_center_size'. This input also can be a 3-D Tensor with shape [N, M, 4] when `code_type` is 'decode_center_size'.

File: python/paddle/fluid/layers/detection.py [hunk 2]
```diff
@@ -776,5 +788,7 @@ def _is_list_or_tuple_and_equal(data, length, err_info):
 
         box = tensor.concat(reshaped_boxes)
         var = tensor.concat(reshaped_vars)
+        mbox_locs_concat = tensor.concat(mbox_locs, axis=1)
+        mbox_confs_concat = tensor.concat(mbox_confs, axis=1)
 
-    return mbox_locs, mbox_confs, box, var
+    return mbox_locs_concat, mbox_confs_concat, box, var
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: Please update the doc.

File: python/paddle/fluid/tests/unittests/test_box_coder_op.py [hunk 3]
```diff
@@ -72,8 +71,14 @@ def batch_box_coder(prior_box, prior_box_var, target_box, lod, code_type):
...
-        box_coder(target_box[lod[i]:lod[i + 1], :], prior_box, prior_box_var,
-                  output_box[lod[i]:lod[i + 1], :, :], code_type)
+        if (code_type == "EncodeCenterSize"):
+            box_coder(target_box[lod[i]:lod[i + 1], :], prior_box,
+                      prior_box_var, output_box[lod[i]:lod[i + 1], :, :],
+                      code_type)
+        elif (code_type == "DecodeCenterSize"):
+            box_coder(target_box[lod[i]:lod[i + 1], :, :], prior_box,
+                      prior_box_var, output_box[lod[i]:lod[i + 1], :, :],
+                      code_type)
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/test_box_coder_op.py]: 当code_type  = `DecodeCenterSize` 时：

从 C++ 实现来看
 https://github.com/PaddlePaddle/Paddle/pull/8647/files#diff-786168ab2d44e839bfa1616a8cd7430fR92

```
    for (int64_t i = 0; i < row; ++i) {
       for (int64_t j = 0; j < col; ++j) {	      
         size_t offset = i * col * len + j * len;
```
这里代码计算offset时target_box当做Tensor算的，但Python这里确实LoDTensor，按照lod访问的。

Comment (Noplz) [reply python/paddle/fluid/tests/unittests/test_box_coder_op.py]: 这个LoD 我给的是1，2，3，4，5其实就是tensor吧，当时为了和之前保持一致就没改

File: python/paddle/fluid/layers/detection.py (hunk 2 above)
Comment (Noplz) [reply python/paddle/fluid/layers/detection.py]: done

File: paddle/fluid/operators/box_coder_op.cc (hunk 1 above)
Comment (Noplz) [reply paddle/fluid/operators/box_coder_op.cc]: done

File: paddle/fluid/operators/box_coder_op.cc [hunk 4]
```diff
@@ -87,8 +96,10 @@ class BoxCoderOpMaker : public framework::OpProtoAndCheckerMaker {
         "OutputBox",
         "(LoDTensor or Tensor) "
         "(Tensor) The output of box_coder_op, a tensor with shape [N, M, 4] "
-        "representing the result of N target boxes encoded/decoded with "
-        "M Prior boxes and variances.");
+        "representing the result of N target boxes encoded with "
+        "M Prior boxes and variances when code_type is 'encode_center_size'. "
```
Comment (qingqing01) [inline paddle/fluid/operators/box_coder_op.cc]: . When code_type is 'encode_center_size',   xxx ？ 把when放到前面，强调一下。后面同样。

File: paddle/fluid/operators/box_coder_op.cc [hunk 5]
```diff
@@ -87,8 +96,10 @@ class BoxCoderOpMaker : public framework::OpProtoAndCheckerMaker {
...
+        "N represents the batchsize and M represents the deocded boxes when "
```
Comment (qingqing01) [inline paddle/fluid/operators/box_coder_op.cc]: batchsize -> batch size
M represents the deocded boxes -> M represents the number of decoded boxes of each instance.

File: paddle/fluid/operators/box_coder_op.cc (hunk 4 above)
Comment (Noplz) [reply paddle/fluid/operators/box_coder_op.cc]: done

File: paddle/fluid/operators/box_coder_op.cc (hunk 5 above)
Comment (Noplz) [reply paddle/fluid/operators/box_coder_op.cc]: done

File: python/paddle/fluid/layers/detection.py [hunk 6]
```diff
@@ -586,12 +586,13 @@ def multi_box_head(inputs,
        name(str): Name of the prior box layer. Default: None.
 
     Returns:
-        mbox_loc(list): The predicted boxes' location of the inputs.
-             The layout of each element is [N, H, W, Priors]. Priors
+        mbox_loc(Variable): The predicted boxes' location of the inputs.
+             The layout of each element is [N, H*W*Priors, 4]. Priors
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: The layout of each element is [N, HWPriors, C]. Priors

->

The layout is [N, HWPriors, C], where Priors xxx

File: python/paddle/fluid/layers/detection.py [hunk 7]
```diff
@@ -586,12 +586,13 @@ def multi_box_head(inputs,
...
              is the number of predicted boxof each position of each input.
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: boxof -> boxes

File: python/paddle/fluid/layers/detection.py [hunk 8]
```diff
@@ -586,12 +586,13 @@ def multi_box_head(inputs,
...
-        mbox_conf(list): The predicted boxes' confidence of the inputs.
-             The layout of each element is [N, H, W, Priors]. Priors
+        mbox_conf(Variable): The predicted boxes' confidence of the inputs.
+             The layout of each element is [N, H*W*Priors, C]. Priors
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: The layout of each element is [N, H*W*Priors, C]. Priors

->

The layout is [N, H*W*Priors, C], where Priors xxx

File: python/paddle/fluid/layers/detection.py (hunk 6 above)
Comment (Noplz) [reply python/paddle/fluid/layers/detection.py]: done

File: python/paddle/fluid/layers/detection.py (hunk 7 above)
Comment (Noplz) [reply python/paddle/fluid/layers/detection.py]: done

File: python/paddle/fluid/layers/detection.py (hunk 8 above)
Comment (Noplz) [reply python/paddle/fluid/layers/detection.py]: done

**Review by qingqing01 — APPROVED** [review]

File: python/paddle/fluid/layers/detection.py [hunk 9]
```diff
@@ -751,7 +752,12 @@ def _is_list_or_tuple_and_equal(data, length, err_info):
             stride=stride)
 
         mbox_loc = nn.transpose(mbox_loc, perm=[0, 2, 3, 1])
-        mbox_locs.append(mbox_loc)
+        new_shape = [
+            mbox_loc.shape[0],
+            mbox_loc.shape[1] * mbox_loc.shape[2] * mbox_loc.shape[3] / 4, 4
```
Comment (velconia) [inline python/paddle/fluid/layers/detection.py:757]: @Noplz 最近在做Python3的迁移, 请问:

```mbox_loc.shape[1] * mbox_loc.shape[2] * mbox_loc.shape[3] / 4```

这段代码的输出是否等价于

```mbox_loc.shape[1] * mbox_loc.shape[2] * mbox_loc.shape[3] // 4```

呢? Python3 除法改动链接在[这里](http://python3porting.com/preparing.html#division-section)

File: python/paddle/fluid/layers/detection.py [hunk 10]
```diff
@@ -762,11 +768,18 @@ def _is_list_or_tuple_and_equal(data, length, err_info):
             padding=pad,
             stride=stride)
         conf_loc = nn.transpose(conf_loc, perm=[0, 2, 3, 1])
-        mbox_confs.append(conf_loc)
+        new_shape = [
+            conf_loc.shape[0], conf_loc.shape[1] * conf_loc.shape[2] *
+            conf_loc.shape[3] / num_classes, num_classes
```
Comment (velconia) [inline python/paddle/fluid/layers/detection.py:773]: 同上

**Pull request merged**